site/src/pages/DeploySettingsPage/OAuth2AppsSettingsPage/EditOAuth2AppPageView.tsx

```diff
@@ -0,0 +1,308 @@
+            />
+
+            <Stack direction="row">
+              <div>
+                <strong>Client ID:</strong>
+              </div>
+              <CopyableValue value={app.id}>{app.id}</CopyableValue>
+            </Stack>
+
+            <Divider css={{ borderColor: theme.palette.divider }} />
+
+            <OAuth2AppForm
+              app={app}
+              onSubmit={updateApp}
+              isUpdating={isUpdating === "update-app"}
+              error={error}
+              actions={
+                <Button
+                  variant="outlined"
+                  type="button"
+                  color="error"
+                  onClick={() => setShowDelete(true)}
+                >
+                  Delete&hellip;
+                </Button>
+              }
+            />
+
+            <Divider css={{ borderColor: theme.palette.divider }} />
+
+            <OAuth2AppSecretsTable
+              secrets={secrets}
+              generateAppSecret={generateAppSecret}
+              deleteAppSecret={deleteAppSecret}
+              isLoadingSecrets={isLoadingSecrets}
+              isUpdating={isUpdating}
+            />
+          </>
+        )}
+      </Stack>
+    </>
+  );
+};
+
+type OAuth2AppSecretsTableProps = {
+  secrets?: TypesGen.OAuth2AppSecret[];
+  generateAppSecret: () => void;
+  isLoadingSecrets: boolean;
+  isUpdating:
+    | "update-app"
+    | "create-secret"
+    | "delete-app"
+    | "delete-secret"
+    | false;
+  deleteAppSecret: (id: string) => void;
+};
+
+const OAuth2AppSecretsTable: FC<OAuth2AppSecretsTableProps> = ({
+  secrets,
+  generateAppSecret,
+  isLoadingSecrets,
+  isUpdating,
+  deleteAppSecret,
+}) => {
+  return (
+    <>
+      <Stack
+        alignItems="baseline"
+        direction="row"
+        justifyContent="space-between"
+      >
+        <h2>Client secrets</h2>
+        <LoadingButton
+          loading={isUpdating === "create-secret"}
+          type="submit"
+          variant="contained"
+          onClick={() => generateAppSecret()}
+        >
+          Generate secret
+        </LoadingButton>
+      </Stack>
+
+      <TableContainer>
+        <Table>
+          <TableHead>
+            <TableRow>
+              <TableCell width="80%">Secret</TableCell>
+              <TableCell width="20%">Last Used</TableCell>
+              <TableCell width="1%" />
+            </TableRow>
+          </TableHead>
+          <TableBody>
+            {isLoadingSecrets && <TableLoader />}
+            {!isLoadingSecrets && (!secrets || secrets?.length === 0) && (
+              <TableRow>
+                <TableCell colSpan={999}>
+                  <div css={{ textAlign: "center" }}>
+                    No client secrets have been generated.
+                  </div>
+                </TableCell>
+              </TableRow>
+            )}
+            {!isLoadingSecrets &&
+              secrets &&
+              secrets?.length > 0 &&
+              secrets?.map((secret) => (
```

Comment:
The second and third boolean checks shouldn't be necessary, because if secrets is undefined, the `?.` will short-circuit the expression and render nothing out

Reply:
Oh right!  Good catch.